components/node-labeler/cmd/run.go

```diff
@@ -249,6 +287,212 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req reconcile.Request) (r
 	return reconcile.Result{}, nil
 }
 
+type NodeScaledownAnnotationController struct {
+	client.Client
+	nodesToReconcile chan string
+	stopChan         chan struct{}
+}
+
+func NewNodeScaledownAnnotationController(client client.Client) (*NodeScaledownAnnotationController, error) {
+	controller := &NodeScaledownAnnotationController{
+		Client:           client,
+		nodesToReconcile: make(chan string, 1000),
+		stopChan:         make(chan struct{}),
+	}
+
+	return controller, nil
+}
+
+func (c *NodeScaledownAnnotationController) SetupWithManager(mgr ctrl.Manager) error {
+	go c.reconciliationWorker()
+	go c.periodicReconciliation()
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("node-scaledown-annotation-controller").
+		For(&workspacev1.Workspace{}).
+		WithEventFilter(c.workspaceFilter()).
+		Complete(c)
+}
+
+// periodicReconciliation periodically reconciles all nodes in the cluster
+func (c *NodeScaledownAnnotationController) periodicReconciliation() {
+	ticker := time.NewTicker(5 * time.Minute)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			log.Info("starting periodic full reconciliation")
+			ctx := context.Background()
+			if _, err := c.reconcileAllNodes(ctx); err != nil {
+				log.WithError(err).Error("periodic reconciliation failed")
+			}
+		case <-c.stopChan:
+			return
```

Comment:
nit: 
Consider logging stop

Reply:
Addressed it in [`1584a52` (#20509)](https://github.com/gitpod-io/gitpod/pull/20509/commits/1584a52583dc2275358ec74943361667743d8844)